abstract/encoding.go

```diff
@@ -292,8 +292,10 @@ func (en *encoder) value(obj interface{}, depth int) error {
 		}
 
 	case reflect.Int:
-		i := int64(v.Int())
-		return binary.Write(en.w, binary.BigEndian, i)
+		t := reflect.TypeOf(int32(0))
```

Comment:
Maybe do this as "var int32Type = ..." in package scope, so the reflect.Typeof call only needs to be done once and not for every int unmarshaled.

Also, why do you want to use reflection (reflect.Convert) instead of just marshaling it from an int32 variable and doing a simple non-reflective test to make sure the int32 variable correctly represents the originally passed int?  Reflection calls are definitely expensive (though I don't have a good sense for just how expensive), and using reflect.Convert for this just seems like one definitely unnecessary reflection call.

Finally, are we sure panic is what we want to happen if the written int is too big?  I see that panic is indeed consistent with the way some other encoding errors are handled in the encoder side, but I'm not sure that's the right way given that the encoding API actually has a graceful error-return path (which might currently never be used but users of this module might understandably *expect* it to be used to report encoding errors).